File: examples/javascript/basic-kv/kv_server.js
#!/usr/bin/env node
const msg = require("../util/msg");

class Server {
  constructor() {
    this.kvstore = new Map();
    this.nodeId = msg.getNodeId();
    this.isLeader = this.nodeId === "kv_server_1"; // server1 is leader
    this.pendingAcks = new Map(); // reqid -> {count, total, clientId, message}
  }

  apply(message) {
    const { key, cmd, value } = message;
    if (cmd === "W") {
      this.kvstore.set(key, value);
    }
    return this.kvstore.get(key);
  }

  quorumSize() {
    // For N servers, need (N/2 + 1) for majority
    const totalServers = msg.getNodes("kv_server.*").length;
    return Math.floor(totalServers / 2) + 1;
  }

  async onAppendReq(message) {
    // Follower handling append request from leader
    this.apply(message);
    msg.send({
      to: message.from,
      from: this.nodeId,
      type: "APPEND_RESP",
      reqid: message.reqid,
    });
  }

  async onClientCommand(message) {
    if (!this.isLeader) {
      // Redirect to leader
      msg.send({
        to: message.from,
        from: this.nodeId,
        type: "CMD_RESP",
        redirect: "kv_server_1",
      });
      return;
    }

    if (message.cmd === "R") {
      // Handle reads directly
      const value = this.kvstore.get(message.key);
      msg.send({
        to: message.from,
        from: this.nodeId,
        type: "CMD_RESP",
        value: value,
        reqid: message.reqid,
      });
    } else if (message.cmd === "W") {
      // Replicate writes
      this.replicate(message);
    }
  }

  replicate(message) {
    // Send append requests to all followers (excluding self)
    const others = msg.siblingNodes().filter((id) => id !== this.nodeId);
    for (const serverId of others) {
      msg.send({
        to: serverId,
        from: this.nodeId,
        type: "APPEND_REQ",
        cmd: message.cmd,
        key: message.key,
        value: message.value,
        reqid: message.reqid,
      });
    }

    // Track pending acks
    this.pendingAcks.set(message.reqid, {
      count: 1, // Include leader's own ack
      total: msg.getNodes("kv_server.*").length,
      clientId: message.from,
      message: message,
    });
  }

  onAppendResp(message) {
    const pending = this.pendingAcks.get(message.reqid);
    if (!pending) return;

    pending.count++;
    console.error(
      `Got ${pending.count} acks out of ${this.quorumSize()} needed`
    );

    // Only respond when we have majority (including leader's own ack)
    if (pending.count >= this.quorumSize()) {
      // Apply to leader's state
      this.apply(pending.message);

      // Respond to client
      msg.send({
        to: pending.clientId,
        from: this.nodeId,
        type: "CMD_RESP",
        success: true,
        reqid: message.reqid,
      });

      this.pendingAcks.delete(message.reqid);
    }
  }
  async listen() {
    while (true) {
      const message = await msg.recv();
      switch (message.type) {
        case "APPEND_REQ":
          await this.onAppendReq(message);
          break;
        case "APPEND_RESP":
          this.onAppendResp(message);
          break;
        case "CMD_REQ":
          await this.onClientCommand(message);
          break;
      }
    }
  }
}

// Start server
const server = new Server();
server.listen().catch(console.error);

File: examples/python/basic-kv/kv_server.py
#!/usr/bin/env python3

from typing import Dict, Tuple
from util.msg import send, recv, node_id

# Message types
CMD_REQ = "CMD_REQ"        # From client to primary
REPLICATE_REQ = "REP_REQ"  # From primary to secondaries
REPLICATE_RESP = "REP_RESP" # From secondaries to primary
CMD_RESP = "CMD_RESP"      # From primary to client
REDIRECT = "REDIRECT"      # From secondary to client

# Store format: key -> (value, version)
kvstore: Dict[str, Tuple[str, int]] = {}

me = node_id()
is_primary = me == "kv_server_1"  # Check if this server is primary

# For primary to track replication responses
pending_writes = {}  # reqid -> (client_id, key, value, responses)

while True:
    msg = recv()
    msg_type = msg["type"]

    if msg_type == CMD_REQ:
        if not is_primary:
            # Secondary servers redirect to primary
            send({
                "to": msg["from"],
                "from": me,
                "type": REDIRECT,
                "primary": "kv_server_1"
            })
            continue

        if msg["cmd"] == "R":
            # Handle reads directly on primary
            key = msg["key"]
            val = kvstore.get(key, None)
            if val:
                value, version = val
                send({
                    "to": msg["from"],
                    "from": me,
                    "type": CMD_RESP,
                    "success": True,
                    "value": value,
                    "version": version
                })
            else:
                send({
                    "to": msg["from"],
                    "from": me,
                    "type": CMD_RESP,
                    "success": False
                })

        elif msg["cmd"] == "W":
            # Forward write to other servers
            reqid = msg["reqid"]
            key = msg["key"]
            value = msg["value"]
            
            # Store write info
            pending_writes[reqid] = (msg["from"], key, value, 0)
            
            # Forward to secondaries
            for server in ["kv_server_2", "kv_server_3"]:
                send({
                    "to": server,
                    "from": me,
                    "type": REPLICATE_REQ,
                    "reqid": reqid,
                    "key": key,
                    "value": value
                })

    elif msg_type == REPLICATE_REQ:
        # Secondary servers handle replication
        key = msg["key"]
        value = msg["value"]
        version = 1
        if key in kvstore:
            (_, old_version) = kvstore[key]
            version = old_version + 1
        kvstore[key] = (value, version)
        
        # Acknowledge to primary
        send({
            "to": msg["from"],
            "from": me,
            "type": REPLICATE_RESP,
            "reqid": msg["reqid"]
        })

    elif msg_type == REPLICATE_RESP:
        # Primary handles replication responses
        reqid = msg["reqid"]
        if reqid in pending_writes:
            client_id, key, value, responses = pending_writes[reqid]
            responses += 1
            
            if responses == 2:  # Both secondaries responded
                # Update primary's state
                version = 1
                if key in kvstore:
                    (_, old_version) = kvstore[key]
                    version = old_version + 1
                kvstore[key] = (value, version)
                
                # Respond to client
                send({
                    "to": client_id,
                    "from": me,
                    "type": CMD_RESP,
                    "success": True,
                    "version": version
                })
                
                # Clean up
                del pending_writes[reqid]
            else:
                # Still waiting for more responses
                pending_writes[reqid] = (client_id, key, value, responses)

    else:
        raise Exception(f"Unknown message type: {msg_type}")